Add Carreau–Yasuda non-Newtonian viscosity model

sherin-joseph opened the pull request:
modified:   src/TrixiParticles.jl
	modified:   src/io/io.jl
	modified:   src/schemes/fluid/viscosity.jl
	modified:   test/schemes/fluid/viscosity.jl

sherin-joseph committed:
Changes to be committed:
	modified:   src/TrixiParticles.jl
	modified:   src/io/io.jl
	modified:   src/schemes/fluid/viscosity.jl
	modified:   test/schemes/fluid/viscosity.jl

sherin-joseph committed:
Changes to be committed:
	modified:   docs/src/refs.bib
	modified:   docs/src/systems/fluid.md
	modified:   src/schemes/fluid/viscosity.jl

sherin-joseph committed:
Merge branch 'main' into non-newtonian

sherin-joseph committed:
Merge branch 'non-newtonian' of https://github.com/sherin-joseph/TrixiParticles.jl into non-newtonian

sherin-joseph committed:
Merge branch 'main' into non-newtonian

sherin-joseph committed:
Merge branch 'main' into non-newtonian

svchb committed:
Merge branch 'main' into non-newtonian

svchb commented on docs/src/systems/fluid.md:
```diff
@@ -155,6 +155,48 @@ The inter-particle-averaged shear stress is defined as:
+originally proposed by [Carreau (1972)](@cite Carreau1972) and extended by
+[Yasuda et al. (1981)](@cite Yasuda1981). In this model, the kinematic viscosity
+depends on the local shear rate. This makes it suitable for shear-thinning and
+shear-thickening fluids, such as polymer solutions or blood-like flows.
```
```suggestion
shear-thickening fluids, such as polymer solutions or blood-like fluids.
```

svchb commented on src/schemes/fluid/viscosity.jl:
```diff
@@ -447,3 +447,98 @@ function kinematic_viscosity(system, viscosity::ViscosityMorrisSGS, smoothing_le
+    m_a, m_b,
+    rho_a, rho_b,
+    grad_kernel,
+)
```
formatting

sherin-joseph committed:
Update docs/src/systems/fluid.md

Co-authored-by: Sven Berger <berger.sven@gmail.com>

sherin-joseph committed:
Merge branch 'trixi-framework:main' into non-newtonian

sherin-joseph committed:
modified:   docs/src/refs.bib
	modified:   docs/src/systems/fluid.md
	modified:   src/schemes/fluid/viscosity.jl
	modified:   test/schemes/fluid/viscosity.jl

sherin-joseph commented:
Ready for re-review.  @svchb

svchb committed:
Merge branch 'main' into non-newtonian

sherin-joseph committed:
Merge branch 'trixi-framework:main' into non-newtonian

efaulhaber commented on docs/src/systems/fluid.md:
```diff
@@ -155,6 +155,48 @@ The inter-particle-averaged shear stress is defined as:
 
+#### ViscosityCarreauYasuda
+
+
```
```suggestion
```

efaulhaber commented on src/schemes/fluid/viscosity.jl:
```diff
@@ -447,3 +447,68 @@ function kinematic_viscosity(system, viscosity::ViscosityMorrisSGS, smoothing_le
+
+Non-Newtonian viscosity model based on the Carreau–Yasuda law [Carreau (1972)](@cite Carreau1972), [Yasuda et al. (1981)](@cite Yasuda1981).
+
+See [`Viscosity`](@ref viscosity_sph) for an overview and comparison of implemented viscosity models.
```
```suggestion
See [the docs on viscosity](@ref viscosity_sph) for an overview and comparison of implemented viscosity models.
```

sherin-joseph committed:
Update docs/src/systems/fluid.md

Co-authored-by: Erik Faulhaber <44124897+efaulhaber@users.noreply.github.com>

sherin-joseph committed:
Update src/io/io.jl

Co-authored-by: Erik Faulhaber <44124897+efaulhaber@users.noreply.github.com>

sherin-joseph committed:
Update src/schemes/fluid/viscosity.jl

Co-authored-by: Erik Faulhaber <44124897+efaulhaber@users.noreply.github.com>

sherin-joseph committed:
new file:   examples/fluid/bubble_rise_carreau_yasuda.jl

sherin-joseph commented:
https://github.com/user-attachments/assets/b6b4ff42-e189-4b74-92f7-ed02ef11f8fb


https://github.com/user-attachments/assets/ab4333fa-8dcd-4dcf-b9cf-b16b29ca7986



<img width="1152" height="864" alt="compare_ar" src="https://github.com/user-attachments/assets/de85a0e4-8466-473b-b89d-25da5dd6ec22" />
<img width="1152" height="864" alt="compare_vel" src="https://github.com/user-attachments/assets/77b31004-c4cd-4065-a36a-5776f150e452" />
<img width="1152" height="864" alt="compare_yc" src="https://github.com/user-attachments/assets/b40ae4bf-2d67-467f-a067-c8fff10498a3" />

For validation I am working with a high resolution version of this example right now. Will update once I have the results.

sherin-joseph committed:
Update comment for bubble_rise_carreau_yasuda.jl

sherin-joseph committed:
Delete examples/fluid/bubble_rise_carreau_yasuda.jl

sherin-joseph committed:
new file:   examples/fluid/bubble_rise.jl

sherin-joseph committed:
Merge branch 'non-newtonian' of https://github.com/sherin-joseph/TrixiParticles.jl into non-newtonian

sherin-joseph committed:
Merge branch 'trixi-framework:main' into non-newtonian

sherin-joseph committed:
Merge branch 'non-newtonian' of https://github.com/sherin-joseph/TrixiParticles.jl into non-newtonian

sherin-joseph committed:
Delete examples/fluid/bubble_rise.jl

svchb committed:
Merge branch 'main' into non-newtonian

Commit:
Format: automatic style check fixes

svchb committed:
Merge branch 'main' into non-newtonian
